src/crimson/os/seastore/transaction_manager.h

```diff
@@ -425,124 +434,90 @@ class TransactionManager : public ExtentCallbackInterface {
-				  ? pin->get_intermediate_key()
-				  : L_ADDR_NULL,
-              original_paddr = pin->get_val(),
-              original_len = pin->get_length()](auto ext) mutable {
-      std::optional<ceph::bufferptr> original_bptr;
+    return seastar::do_with(
+      std::vector<LogicalCachedExtentRef>(),
+      std::move(pin),
+      std::move(remaps),
+      [&t, this](auto &extents, auto &pin, auto &remaps) {
+      laddr_t original_laddr = pin->get_key();
+      extent_len_t original_len = pin->get_length();
+      paddr_t original_paddr = pin->get_val();
       LOG_PREFIX(TransactionManager::remap_pin);
       SUBDEBUGT(seastore_tm,
-        "original laddr: {}, original paddr: {}, original length: {},"
-	" intermediate_base: {}, intermediate_key: {},"
-        " remap to {} extents",
-        t, original_laddr, original_paddr, original_len,
-	intermediate_base, intermediate_key, remaps.size());
-      ceph_assert(
-	(intermediate_base == L_ADDR_NULL)
-	  == (intermediate_key == L_ADDR_NULL));
-      ceph_assert(full_extent_integrity_check
-	  ? (ext && ext->is_fully_loaded())
-	  : true);
-      if (ext) {
-        ceph_assert(!ext->is_mutable());
-        ceph_assert(ext->get_length() >= original_len);
-	ceph_assert(ext->get_paddr() == original_paddr);
-        original_bptr = ext->get_bptr();
-      }
-      return seastar::do_with(
-        std::array<LBAMappingRef, N>(),
-        0,
-        std::move(original_bptr),
-        std::vector<remap_entry>(remaps.begin(), remaps.end()),
-        [this, &t, original_laddr, original_paddr,
-	original_len, intermediate_base, intermediate_key]
-        (auto &ret, auto &count, auto &original_bptr, auto &remaps) {
-        return _dec_ref(t, original_laddr, false
-        ).si_then([this, &t, &original_bptr, &ret, &count,
-		   &remaps, intermediate_base, intermediate_key,
-                   original_laddr, original_paddr, original_len](auto) {
-          return trans_intr::do_for_each(
-            remaps.begin(),
-            remaps.end(),
-            [this, &t, &original_bptr, &ret,
-	    &count, intermediate_base, intermediate_key,
-	    original_laddr, original_paddr, original_len](auto &remap) {
-            LOG_PREFIX(TransactionManager::remap_pin);
-            auto remap_offset = remap.offset;
-            auto remap_len = remap.len;
-            auto remap_laddr = original_laddr + remap_offset;
-            auto remap_paddr = original_paddr.add_offset(remap_offset);
-	    if (intermediate_key != L_ADDR_NULL) {
-	      remap_paddr = original_paddr;
+	"original laddr: {}, original paddr: {}, original length: {},"
+	" remap to {} extents",
+	t, original_laddr, original_paddr, original_len, remaps.size());
+      // The according extent might be stable or pending.
+      auto fut = base_iertr::now();
+      if (!pin->is_indirect()) {
+	auto fut2 = base_iertr::make_ready_future<TCachedExtentRef<T>>();
+	if (full_extent_integrity_check) {
+	  fut2 = read_pin<T>(t, pin->duplicate());
```

Comment:
I think [your above comment](https://github.com/ceph/ceph/pull/57262#discussion_r1602533286) is an example of why keeping pin an `unique_ptr`. A pin may get invalidated once its parent lba leaf node is modified, mandating `duplicating a pin` explicit can, to some extent, notifying developers of the danger of using an invalid pin.

I think we can make it shared once we have implemented `LBAMapping::is_valid()`, what do you think?

Reply:
(extended discussion)

For remap_pin(), it is fine to require an rvalue pin to clarify the intention that the original pin will become invalid.
For read_pin(), it seems inconvenient to require an rvalue pin as the pin is still valid after reading (e.g. looks object data handler wants to keep track the pin)

> A pin may get invalidated once its parent lba leaf node is modified, mandating duplicating a pin explicit can, to some extent, remind developers of the danger of using an invalid pin.

That's true, and this defines how object data handler should work -- that it should not modify lba tree and reading a pin concurrently in the same transaction.

However, the method `duplicate()` looks to be an overkill because it involves an extra allocation, so it is not merely a reminder.

> I think we can make it shared once we have implemented LBAMapping::is_valid(), what do you think?

I think keeping the pin as unique-ptr is also fine, so that read-pin can become `read_pin(Transaction &, LBAMapping &)` -- user needs to make sure the pin outlives reading.

Anyway, for this PR, I suppose we can keep the style of `remap_pin(std::move(pin) ... )`, what do you think?